CodeReviewStandards/rfc-008-review-checklist.md

```diff
@@ -0,0 +1,46 @@
+# Pull Request Standards
+
+## Summary
+
+A checklist designed to help facilitate code reviews by outlining a pre-review checklist to follow for the raiser of the pull request, and a reviewer checklist.
+
+## Problem
+
+Pull requests follow no set format and vary between projects, which in turn causes reviews to take longer and require more context switching. Code review and commentary often happens 'offline', which results in reduced context and fewer points of reference when reviewing changes later.
+
+## Proposal
+
+Before raising a pull request, the pull request SHOULD meet the following criteria:
+
+ - [ ] Pull request description outlines _what has changed_
+ - [ ] Pull request description outlines _why the change was made_
+ - [ ] If the project has a linter, it SHOULD not report any issues with the pull request
+ - [ ] Code follows project specific coding conventions
+ - [ ] Code follows language specific conventions in the absence of a project specific convention
+ - [ ] Code has been refactored with quality and readability in mind
+ - [ ] Tests are obviously the product of practising TDD (where appropriate)
+ - [ ] There are high level feature/acceptance tests in place as appropriate
+ - [ ] Tests have been refactored - aim for quality, readability and maintainability
+ - [ ] Test suite shows that the change has been tested to adequate confidence.
+
+When reviewing a pull request, reviewer SHOULD consider at least the following questions:
+
+ - [ ] Does the code achieve the goal outlined in _why the change was made_?
+ - [ ] Could the change have been implemented using a different strategy that would be a credible alternative? If so, why would the alternative be better?
+ - [ ] Does the code follow all of the testing and standards requirements in the pre-review checklist?
+ - [ ] Does the code introduce unnecessary additional complexity or lengthy test cycles?
+ - [ ] Is the code readable?
+ - [ ] Is the code maintainable?
```

Comment:
What defines maintainable / readable :evil: ?

Reply:
You don't have to be able to say with 100% confidence that the code is forever maintainable. It's a question to ask yourself as a reviewer, is there anything I could change to make this code more maintainable? - Is it worth rewording to this extent?

Reply:
Yeah I'm largely in favour of this PR.

Just wondering how far these guidelines should go, i.e. at what point do we give up and say "actually, just use your best judgement, we trust you" vs "no you look 4 these things pls"